File: internal/cassette/privacy.go
package cassette

import (
	"encoding/json"
	"fmt"
	"math"
	"reflect"
	"regexp"
	"strings"
	"unicode"
)

type PrivacyMode string

const (
	PrivacySafe      PrivacyMode = "safe"
	PrivacyHideAll   PrivacyMode = "hide_all"
	PrivacyTransform PrivacyMode = "transform"
)

type PrivacyOptions struct {
	Mode      PrivacyMode
	Sanitizer func(any) any
}

const redactedValue = "[REDACTED]"

var hiddenPayload = map[string]any{"value_hidden": true}

var safeEnvelopeFields = map[string]struct{}{
	"schema_version": {},
	"event":          {},
	"trace_id":       {},
	"span_id":        {},
	"name":           {},
	"provider":       {},
	"model":          {},
	"status":         {},
	"latency_ms":     {},
}

var sensitiveExactKeys = map[string]struct{}{
	"api_key":           {},
	"api_token":         {},
	"authorization":     {},
	"bearer":            {},
	"client_secret":     {},
	"cookie":            {},
	"cookies":           {},
	"csrf_token":        {},
	"id_token":          {},
	"password":          {},
	"passwd":            {},
	"private_key":       {},
	"pwd":               {},
	"refresh_token":     {},
	"secret":            {},
	"secret_access_key": {},
	"secret_key":        {},
	"session_cookie":    {},
	"session_token":     {},
	"set_cookie":        {},
	"token":             {},
}

var sensitiveKeyParts = []string{
	"access_key",
	"api_key",
	"api_token",
	"authorization",
	"client_secret",
	"cookie",
	"csrf_token",
	"credential",
	"id_token",
	"password",
	"private_key",
	"refresh_token",
	"secret_access_key",
	"secret_key",
	"session_token",
}

var secretPatterns = []*regexp.Regexp{
	regexp.MustCompile(`(?s)-----BEGIN (?:[A-Z0-9 ]*PRIVATE KEY|OPENSSH PRIVATE KEY)-----.*?-----END (?:[A-Z0-9 ]*PRIVATE KEY|OPENSSH PRIVATE KEY)-----`),
	regexp.MustCompile(`(?i)\b(?:set-cookie|cookie)\s*:\s*[^,\n]+`),
	regexp.MustCompile(`(?i)\b[^=\s;]+=[^;\s]+;\s*(?:HttpOnly|Secure|SameSite=[A-Za-z]+)`),
	regexp.MustCompile(`(?i)\b[a-z][a-z0-9+.-]*://[^\s/@:]+:[^\s/@]+@[^\s]+`),
	regexp.MustCompile(`(?i)(?m)(^|[^A-Za-z0-9_])["']?(?:aws_secret_access_key|secret_access_key|access_key_secret|client_secret|api_key|api_token|access_token|refresh_token|id_token|session_token|csrf_token|token|password|passwd|pwd)["']?\s*[:=]\s*(?:"[^"]*"|'[^']*'|[^,\s;&}\]]+)`),
	regexp.MustCompile(`\b(?:AKIA|ASIA|AIDA|AROA|AGPA|AIPA|ANPA)[A-Z0-9]{16}\b`),
	regexp.MustCompile(`sk-[A-Za-z0-9_-]{8,}`),
	regexp.MustCompile(`(?i)Bearer\s+[A-Za-z0-9._-]+`),
}

func sanitizeEvent(fields map[string]any, options PrivacyOptions) (map[string]any, error) {
	mode := options.Mode
	if mode == "" {
		mode = PrivacySafe
	}

	switch mode {
	case PrivacySafe:
		sanitized, ok := sanitizeValue(fields).(map[string]any)
		if !ok {
			return nil, fmt.Errorf("sanitize cassette event: expected object")
		}
		return sanitized, nil
	case PrivacyTransform:
		if options.Sanitizer == nil {
			return nil, fmt.Errorf("privacy mode %q requires a sanitizer", PrivacyTransform)
		}
		sanitized, ok := sanitizeValue(options.Sanitizer(fields)).(map[string]any)
		if !ok {
			return nil, fmt.Errorf("sanitize cassette event: expected object")
		}
		return sanitized, nil
	case PrivacyHideAll:
		return hideEventPayload(fields), nil
	default:
		return nil, fmt.Errorf("unknown privacy mode %q", mode)
	}
}

func hideEventPayload(fields map[string]any) map[string]any {
	hidden := map[string]any{}
	for key, value := range fields {
		if _, ok := safeEnvelopeFields[key]; !ok {
			continue
		}
		hidden[key] = sanitizeValue(value)
	}

	eventType, _ := fields["event"].(string)
	switch eventType {
	case "llm.call":
		hidden["input_hash"] = "hidden:payload"
	case "llm.response", "tool.response":
		if _, ok := fields["error"]; ok {
			hidden["error"] = "[HIDDEN]"
		} else {
			hidden["output"] = cloneMap(hiddenPayload)
		}
	case "agent.step":
		hidden["output"] = cloneMap(hiddenPayload)
	case "retrieval.call":
		hidden["query"] = "[HIDDEN]"
	case "retrieval.response":
		hidden["documents"] = []any{}
	case "error":
		hidden["message"] = "[HIDDEN]"
	}
	return hidden
}

func sanitizeValue(value any) any {
	switch typed := value.(type) {
	case nil:
		return nil
	case string:
		return redactText(typed)
	case bool:
		return typed
	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, json.Number:
		return typed
	case float32:
		if math.IsInf(float64(typed), 0) || math.IsNaN(float64(typed)) {
			return unsupportedPayloadMarker()
		}
		return typed
	case float64:
		if math.IsInf(typed, 0) || math.IsNaN(typed) {
			return unsupportedPayloadMarker()
		}
		return typed
	case map[string]any:
		return sanitizeStringMap(typed)
	case map[any]any:
		result := map[string]any{}
		for key, item := range typed {
			keyText := fmt.Sprint(key)
			if shouldDropKey(keyText) || redactText(keyText) != keyText {
				continue
			}
			result[keyText] = sanitizeValue(item)
		}
		return result
	case []any:
		result := make([]any, 0, len(typed))
		for _, item := range typed {
			result = append(result, sanitizeValue(item))
		}
		return result
	}

	rv := reflect.ValueOf(value)
	if rv.IsValid() && rv.Kind() == reflect.Map {
		result := map[string]any{}
		iter := rv.MapRange()
		for iter.Next() {
			key := iter.Key()
			if !key.CanInterface() {
				return unsupportedPayloadMarker()
			}
			keyText, ok := mapKeyText(key.Interface())
			if !ok {
				return unsupportedPayloadMarker()
			}
			if shouldDropKey(keyText) || redactText(keyText) != keyText {
				continue
			}
			value := iter.Value()
			if !value.CanInterface() {
				return unsupportedPayloadMarker()
			}
			result[keyText] = sanitizeValue(value.Interface())
		}
		return result
	}
	if rv.IsValid() && (rv.Kind() == reflect.Slice || rv.Kind() == reflect.Array) {
		result := make([]any, 0, rv.Len())
		for i := 0; i < rv.Len(); i++ {
			result = append(result, sanitizeValue(rv.Index(i).Interface()))
		}
		return result
	}
	return unsupportedPayloadMarker()
}

func mapKeyText(key any) (string, bool) {
	switch typed := key.(type) {
	case string:
		return typed, true
	case fmt.Stringer:
		return typed.String(), true
	default:
		return "", false
	}
}

func sanitizeStringMap(value map[string]any) map[string]any {
	result := map[string]any{}
	for key, item := range value {
		if shouldDropKey(key) || redactText(key) != key {
			continue
		}
		result[key] = sanitizeValue(item)
	}
	return result
}

func shouldDropKey(key string) bool {
	normalized := normalizeKey(key)
	if _, ok := sensitiveExactKeys[normalized]; ok {
		return true
	}
	if strings.HasPrefix(normalized, "secret_") ||
		strings.HasPrefix(normalized, "token_") {
		return true
	}
	if strings.HasSuffix(normalized, "_password") ||
		strings.HasSuffix(normalized, "_secret") ||
		strings.HasSuffix(normalized, "_token") {
		return true
	}
	for _, part := range sensitiveKeyParts {
		if strings.Contains(normalized, part) {
			return true
		}
	}
	return false
}

func normalizeKey(key string) string {
	runes := []rune(key)
	var builder strings.Builder
	for index, current := range runes {
		if index > 0 && shouldInsertKeySeparator(runes, index) {
			builder.WriteByte('_')
		}
		if unicode.IsLetter(current) || unicode.IsDigit(current) {
			builder.WriteRune(unicode.ToLower(current))
		} else {
			builder.WriteByte('_')
		}
	}
	parts := strings.FieldsFunc(builder.String(), func(r rune) bool { return r == '_' })
	return strings.Join(parts, "_")
}

func shouldInsertKeySeparator(runes []rune, index int) bool {
	current := runes[index]
	previous := runes[index-1]
	if !unicode.IsUpper(current) {
		return false
	}
	if unicode.IsLower(previous) || unicode.IsDigit(previous) {
		return true
	}
	if unicode.IsUpper(previous) && index+1 < len(runes) && unicode.IsLower(runes[index+1]) {
		return true
	}
	return false
}

func redactText(text string) string {
	redacted := text
	for _, pattern := range secretPatterns {
		redacted = pattern.ReplaceAllString(redacted, redactedValue)
	}
	return redacted
}

func unsupportedPayloadMarker() map[string]any {
	return map[string]any{
		"value_unavailable": true,
		"reason":            "unsupported_type",
	}
}

func cloneMap(value map[string]any) map[string]any {
	result := make(map[string]any, len(value))
	for key, item := range value {
		result[key] = item
	}
	return result
}

File: python/agentreplay/privacy.py
"""Payload privacy helpers for cassette recording."""

from __future__ import annotations

import math
import re
from collections.abc import Callable, Mapping
from dataclasses import asdict, is_dataclass
from typing import Any, Literal

PrivacyMode = Literal["safe", "hide_all", "transform"]

REDACTED = "[REDACTED]"
HIDDEN = {"value_hidden": True}

_OMIT = object()

_SAFE_ENVELOPE_FIELDS = {
    "schema_version",
    "event",
    "trace_id",
    "span_id",
    "name",
    "provider",
    "model",
    "status",
    "latency_ms",
}

_SENSITIVE_EXACT_KEYS = {
    "api_key",
    "api_token",
    "authorization",
    "bearer",
    "client_secret",
    "cookie",
    "cookies",
    "csrf_token",
    "id_token",
    "password",
    "passwd",
    "private_key",
    "pwd",
    "refresh_token",
    "secret",
    "secret_access_key",
    "secret_key",
    "session_cookie",
    "session_token",
    "set_cookie",
    "token",
}

_SENSITIVE_KEY_PARTS = (
    "access_key",
    "api_key",
    "api_token",
    "authorization",
    "client_secret",
    "cookie",
    "csrf_token",
    "credential",
    "id_token",
    "password",
    "private_key",
    "refresh_token",
    "secret_access_key",
    "secret_key",
    "session_token",
)

_SECRET_PATTERNS = (
    re.compile(
        r"-----BEGIN (?:[A-Z0-9 ]*PRIVATE KEY|OPENSSH PRIVATE KEY)-----.*?-----END (?:[A-Z0-9 ]*PRIVATE KEY|OPENSSH PRIVATE KEY)-----",
        re.DOTALL,
    ),
    re.compile(r"\b(?:set-cookie|cookie)\s*:\s*[^,\n]+", re.IGNORECASE),
    re.compile(r"\b[^=\s;]+=[^;\s]+;\s*(?:HttpOnly|Secure|SameSite=[A-Za-z]+)", re.IGNORECASE),
    re.compile(r"\b[a-z][a-z0-9+.-]*://[^\s/@:]+:[^\s/@]+@[^\s]+", re.IGNORECASE),
    re.compile(
        r"(?<![A-Za-z0-9_])[\"']?(?:aws_secret_access_key|secret_access_key|access_key_secret|client_secret|api_key|api_token|access_token|refresh_token|id_token|session_token|csrf_token|token|password|passwd|pwd)[\"']?\s*[:=]\s*(?:[\"'][^\"']*[\"']|[^,\s;&}\]]+)",
        re.IGNORECASE,
    ),
    re.compile(r"\b(?:AKIA|ASIA|AIDA|AROA|AGPA|AIPA|ANPA)[A-Z0-9]{16}\b"),
    re.compile(r"sk-[A-Za-z0-9_-]{8,}"),
    re.compile(r"Bearer\s+[A-Za-z0-9._-]+", re.IGNORECASE),
)


def sanitize_event(
    event: Mapping[str, Any],
    *,
    mode: PrivacyMode = "safe",
    sanitizer: Callable[[Any], Any] | None = None,
) -> dict[str, Any]:
    """Return the event payload that is safe to serialize."""

    if mode == "safe":
        return sanitize_payload(event)
    if mode == "transform":
        if sanitizer is None:
            raise ValueError("privacy mode 'transform' requires a sanitizer")
        transformed = sanitizer(event)
        return sanitize_payload(transformed)
    if mode == "hide_all":
        return hide_event_payload(event)
    raise ValueError(f"unknown privacy mode {mode!r}")


def sanitize_payload(value: Any) -> Any:
    """Recursively drop secret keys, redact secret-like strings, and JSON-normalize."""

    payload = _sanitize_value(value)
    if payload is _OMIT:
        return unsupported_payload_marker()
    return payload


def hide_event_payload(event: Mapping[str, Any]) -> dict[str, Any]:
    hidden: dict[str, Any] = {}
    event_type = event.get("event")
    for key, value in event.items():
        text_key = str(key)
        if text_key in _SAFE_ENVELOPE_FIELDS:
            safe_value = _sanitize_value(value)
            if safe_value is not _OMIT:
                hidden[text_key] = safe_value

    if event_type == "llm.call":
        hidden.setdefault("input_hash", "hidden:payload")
    elif event_type in {"llm.response", "tool.response"} and "error" in event:
        hidden["error"] = "[HIDDEN]"
    elif event_type in {"llm.response", "tool.response", "agent.step"}:
        hidden["output"] = dict(HIDDEN)
    elif event_type == "retrieval.call":
        hidden["query"] = "[HIDDEN]"
    elif event_type == "retrieval.response":
        hidden["documents"] = []
    elif event_type == "error":
        hidden["message"] = "[HIDDEN]"

    return hidden


def should_drop_key(key: str) -> bool:
    normalized = normalize_key(key)
    if normalized in _SENSITIVE_EXACT_KEYS:
        return True
    if normalized.startswith(("secret_", "token_")):
        return True
    if normalized.endswith(("_password", "_secret", "_token")):
        return True
    return any(part in normalized for part in _SENSITIVE_KEY_PARTS)


def normalize_key(key: str) -> str:
    text = re.sub(r"([a-z0-9])([A-Z])", r"\1_\2", key)
    text = re.sub(r"([A-Z]+)([A-Z][a-z])", r"\1_\2", text)
    return re.sub(r"[^a-z0-9]+", "_", text.lower()).strip("_")


def redact_text(text: str) -> str:
    redacted = text
    for pattern in _SECRET_PATTERNS:
        redacted = pattern.sub(REDACTED, redacted)
    return redacted


def unsupported_payload_marker() -> dict[str, Any]:
    return {
        "value_unavailable": True,
        "reason": "unsupported_type",
    }


def _sanitize_value(value: Any) -> Any:
    if value is _OMIT:
        return _OMIT
    if value is None:
        return None
    if isinstance(value, str):
        return redact_text(value)
    if isinstance(value, bool) or isinstance(value, int):
        return value
    if isinstance(value, float):
        return value if math.isfinite(value) else unsupported_payload_marker()
    if isinstance(value, Mapping):
        return _sanitize_mapping(value)
    if isinstance(value, (list, tuple)):
        return [_sanitize_item(item) for item in value]
    if is_dataclass(value) and not isinstance(value, type):
        return _sanitize_value(asdict(value))

    model_dump = getattr(value, "model_dump", None)
    if callable(model_dump):
        try:
            return _sanitize_value(model_dump(mode="json", exclude_none=True))
        except TypeError:
            try:
                return _sanitize_value(model_dump())
            except Exception:
                return unsupported_payload_marker()
        except Exception:
            return unsupported_payload_marker()

    to_dict = getattr(value, "to_dict", None)
    if callable(to_dict):
        try:
            return _sanitize_value(to_dict())
        except Exception:
            return unsupported_payload_marker()

    return unsupported_payload_marker()


def _sanitize_mapping(value: Mapping[Any, Any]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, item in value.items():
        key_text = str(key)
        if should_drop_key(key_text) or redact_text(key_text) != key_text:
            continue
        payload = _sanitize_value(item)
        if payload is not _OMIT:
            result[key_text] = payload
    return result


def _sanitize_item(value: Any) -> Any:
    payload = _sanitize_value(value)
    if payload is _OMIT:
        return unsupported_payload_marker()
    return payload
